Added NumbaLinker

kc611 opened the pull request:
Adds a `NumbaLinker` based upon the new `JitLinker` class.

Based upon machinery from https://github.com/pymc-devs/aesara/pull/365 

Cc @brandonwillard

brandonwillard committed:
Create flat functions via AST for JAXification of FunctionGraphs

brandonwillard committed:
Rename modules jax_linker to linker and jax_dispatch to dispatch

brandonwillard committed:
Use jax.numpy.vectorize for Elemwise Composite Ops

brandonwillard committed:
Generalize FunctionGraph conversion function with aesara.link.utils.fgraph_to_python

brandonwillard committed:
Create a generalized JITLinker

kc611 committed:
Added NumbaLinker

kc611 committed:
Generalized Numba conversion of ScalarOps

brandonwillard reviewed (commented):
Well, it looks like you did it!  That's about all it should take.

brandonwillard commented on aesara/link/numba/dispatch.py:
```diff
@@ -0,0 +1,76 @@
+
+    @numba.njit
+    def scalar_func(*args):
+        result = args[0]
+        for arg in args[1:]:
+            result = numpy_func(arg, result)
```
Is this an instance in which the Numba conversion was having trouble dealing with varargs (e.g. `numpy_func(*args)`)?

kc611 replied:
That error occurs when we put a `numba.vectorize` tag on a function which (or it's child functions) has varargs like `*args` . It's the `TODO` listed in `Elemwise` 's Numba conversion.

brandonwillard committed:
Add basic Subtensor support for Numba

`AdvancedSubtensor` will call out to Python (i.e. Numba's object mode), because
that's not currently supported in nopython mode.  This causes problems with
indices that contain slices, because Numba has its own internal representation
of slices and no unboxing for that (i.e. conversion back to Python).

kc611 commented:
Continued in https://github.com/pymc-devs/aesara/pull/372
